nixos/modules/services/abstract/instance/generic.nix

```diff
@@ -0,0 +1,67 @@
+# This corresponds to the arguments of createManagedProcess in svanderburg's proposal
+{ lib, config, options, ... }:
+let
+  inherit (lib) literalExpression mkOption types;
+  pathOrStr = types.coercedTo types.path (x: "${x}") types.str;
+in
+{
+  options = {
+    enable = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Whether to enable the service. A service is enabled by default, because you took the effort to define its attribute.
+      '';
+    };
+    process = mkOption {
+      type = pathOrStr;
+      description = ''
+        When this property is specified, it translates both to: {option}`foreground.process` and {option}`daemon.process`.
+      '';
+    };
+    foreground = {
+      process = lib.mkOption {
+        type  = pathOrStr;
+        description = ''
+          Path to an executable that launches a process in foreground mode.
+        '';
+        default = config.process;
+        defaultText = literalExpression "config.process";
+      };
+      args = lib.mkOption {
+        type = types.listOf pathOrStr;
+        description = ''
+          Arguments to pass to the {option}`foreground.process`.
+        '';
+        default = config.args;
+        defaultText = literalExpression "config.args";
+      };
+    };
+    daemon = {
+      process = lib.mkOption {
+        type = pathOrStr;
+        description = ''
+          Path to an executable that launches a process in daemon mode.
+        '';
+        default = config.process;
+        defaultText = literalExpression "config.process";
+      };
+      args = lib.mkOption {
+        type = types.listOf pathOrStr;
+        description = ''
+          Arguments to pass to the {option}`daemon.process`.
+        '';
+        default = config.args;
+        defaultText = literalExpression "config.args";
+      };
+    };
```

Comment:
I think allowing programs to daemonize themselves will be a trap for this. Foremost new programs don't implement this anymore and a generally discouraged to do so. Old programs implement this in completely varying degrees ranging from full signal handling, creating pid files and being their own process manager to basically doing `program &` and running or anything in between. Also sometimes config options are involved.

IMO we should only take foreground programs and let the service manager handle all of the "daemon" parts. If the service manager doesn't support this, then IMO it is unsuitable.